src/components/ContactModal.test.tsx

```diff
@@ -0,0 +1,233 @@
+import { UpsertContactModal } from "./ContactModal";
+import { contact1, contact2 } from "../mocks/contacts";
+import { mockImplicitAddress } from "../mocks/factories";
+import { fireEvent, render, screen, waitFor } from "../mocks/testUtils";
+import { contactsActions } from "../utils/redux/slices/contactsSlice";
+import { store } from "../utils/redux/store";
+
+describe("UpsertContactModal", () => {
+  describe("on adding contact", () => {
+    describe.each([
+      { testCase: "new contact", renderComponent: () => render(<UpsertContactModal />) },
+      {
+        testCase: "pre-set contact",
+        renderComponent: () =>
+          render(
+            <UpsertContactModal
+              contact={{
+                name: "",
+                pkh: mockImplicitAddress(0).pkh,
+              }}
+            />
+          ),
+      },
+    ])("for $testCase", ({ renderComponent }) => {
+      it("shows correct title & button label for new contact", () => {
+        renderComponent();
+
+        expect(screen.getByRole("dialog")).toHaveTextContent("Add Contact");
+        expect(screen.getByTestId("confirmation-button")).toHaveTextContent("Add to Address Book");
+      });
+
+      it("has editable address & name fields", () => {
+        renderComponent();
+
+        expect(screen.getByLabelText("Address")).toBeEnabled();
+        expect(screen.getByLabelText("Name")).toBeEnabled();
+      });
+
+      it("validates updated address", async () => {
+        renderComponent();
+
+        const addressInput = screen.getByLabelText("Address");
+        fireEvent.change(addressInput, {
+          target: { value: "invalid pkh" },
+        });
+        fireEvent.blur(addressInput);
+
+        await waitFor(() =>
+          expect(screen.getByTestId("address-error")).toHaveTextContent("Invalid address")
+        );
+      });
+
+      it("checks the name is unique", async () => {
+        store.dispatch(contactsActions.upsert(contact2));
+        renderComponent();
+
+        fireEvent.change(screen.getByLabelText("Name"), {
+          target: { value: contact2.name },
+        });
+        fireEvent.blur(screen.getByLabelText("Name"));
+
+        await waitFor(() =>
+          expect(screen.getByTestId("name-error")).toHaveTextContent(
+            "Name must be unique across all accounts and contacts"
+          )
+        );
+      });
+
+      it("adds contact to address book", async () => {
+        store.dispatch(contactsActions.upsert(contact2));
+        renderComponent();
+
+        // Set name
+        fireEvent.change(screen.getByLabelText("Name"), {
+          target: { value: "Test Contact" },
+        });
+        fireEvent.blur(screen.getByLabelText("Name"));
```

Comment:
do you need the `blur` event here? if not maybe you could use `userEvent` instead for writing the text into the input?

Reply:
It doesn't work with just `user.type`. For some reason `user.type` doesn't trigger whatever `blur` triggers.
But I managed to make it work with clicking outside of the text input

Reply:
if you need the blur even then please stick to the `fireEvent`